apache/spark
# [SPARK-26321][SQL] Improve the behavior of sql text splitting for the spark-sql command line

## What changes were proposed in this pull request?

Improve the behavior of sql text splitting for the spark-sql command line.

Currently, the spark-sql command line does not split the sql text correctly, for example, the `;` in double quotes.


## How was this patch tested?
Manually tested:
```
$ ./build/mvn -Phive-thriftserver -DskipTests package
$ bin/spark-sql
> select "^;^";
```

The expected result is `^;^` . However, Spark master will split the SQL by `;`. As as result, `select "^` is not a valid SQL.

```
spark-sql> select "\";";
";
spark-sql> select "\';";
';
```

Unit Tests:
```
$ build/sbt -Phive-thriftserver
> project hive-thriftserver
> testOnly *SparkSQLCLIDriverSuite
> project catalyst
> testOnly *StringUtilsSuite
```

## TODO
Polish the code on SignalHandler. However, it should be another PR.

+ [x] Fix the failed tests.
+ [x] Polish StringUtils.split and add more unit tests
+ [x] Add scaladoc
+ [x] Handling comments
